ykcompile/src/lib.rs

```diff
@@ -175,18 +175,59 @@ impl<TT> TraceCompiler<TT> {
 
     fn place_to_location(&mut self, p: &Place) -> Result<Location, CompileError> {
         if !p.projection.is_empty() {
+            if p.projection.len() > 1 {
+                todo!("chains of projections");
+            }
+
             if Some(p.local) == self.trace_inputs_local {
-                match &p.projection[0] {
-                    Projection::Field(idx) => Ok(Location::new_mem(RDI.code(), (idx * 8) as i32)),
-                    Projection::Deref => unreachable!(),
-                    Projection::Unimplemented(s) => {
-                        Err(CompileError::Unimplemented(format!("{}", s)))
+                let tuple_ty = self.place_ty(&Place::from(p.local));
+                if let Ty::Tuple(tty) = tuple_ty {
+                    match &p.projection[0] {
+                        Projection::Field(idx) => {
+                            let offs = tty.fields.offsets[usize::try_from(*idx).unwrap()];
+                            Ok(Location::new_mem(RDI.code(), i32::try_from(offs).unwrap()))
+                        }
+                        Projection::Deref => unreachable!(),
+                        Projection::Unimplemented(s) => {
+                            Err(CompileError::Unimplemented(format!("{}", s)))
+                        }
                     }
+                } else {
+                    unreachable!(); // Trace inputs are always a tuple.
                 }
             } else {
-                // TODO deal with remaining projections
+                // FIXME deal with remaining projections and chains of projections.
+                let base_ty = self.place_ty(&Place::from(p.local)).clone();
                 match &p.projection[0] {
-                    Projection::Field(0) => self.local_to_location(p.local),
+                    Projection::Field(idx) => match base_ty {
+                        Ty::Struct(sty) => {
```

Comment:
The code for `Struct` and `Tuple` looks identical to me. Is there a reason we don't combine the two, e.g. `Struct | Tuple => {`?

Reply:
I was thinking about this too.

I think there may be scenarios later where we might like the distinction, but I was thinking of making the innards common (in another PR).

Reply:
Should we combine them for now and copy them when/if we actually have to?

Reply:
Ah, you mean this piece of code specifically.

we can't because `tty` and `sty` are different types.

The change I was proposing would make `tty` and `sty` the same type, but it'd be a larger change.

Reply:
Oh they are different types! That wasn't clear from the code. In that case this is for a future PR.